### PR TITLE
Add callbacks to add or keep section number

### PR DESCRIPTION
**Pull request by tuanngocnguyen**
Hi @Syxton ,

I have created a PR for this issue:

https://github.com/Syxton/moodle-block_massaction/issues/115

I have added 2 callbacks for course format to determine if a user can select "Add new section"/"Keep original section"

Would you please have a look and let me know your opinion?

Thanks

### CODE REVIEW

File: classes/form/section_select_form.php [hunk 1]
```diff
@@ -45,6 +45,8 @@
  * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
 class section_select_form extends moodleform {
+    /** Int array of allowed values */
+    protected $allowedvalues = [];
```
Comment (TomoTsuyuki) [inline classes/form/section_select_form.php:49]: the variable name does not really mean the values.
maybe $targetsections or something is better?

File: classes/form/section_select_form.php [hunk 2]
```diff
@@ -100,47 +102,79 @@ public function definition() {
...
     }
+
+    /**
+     * Check if the option should be allowed.
+     * @param bool $allowcondition the condition to allow the option
+     * @param int $optionnumber the option number
+     * @param array $attributes the attributes
+     * @return void
+     */
+    private function is_allowed(bool $allowcondition, int $optionnumber, array &$attributes): void {
```
Comment (TomoTsuyuki) [inline classes/form/section_select_form.php:156]: this function name sounds like for checking value and return bool if condition is ok.
maybe `set_radioption` or something?

File: classes/massactionutils.php [hunk 3]
```diff
@@ -165,6 +165,26 @@ public static function duplicate_cm_to_course(object $course, object $cm): int {
...
 
+    /**
+     * Run callback function from a course format plugin.
+     *
+     * @param int $courseid the target course id
+     * @param string $format the course format
+     * @param string $callbackname the name of the callback function
+     * @return mixed|null none determined return type
+     */
+    private static function run_course_format_callback(int $courseid, string $format, string$callbackname) {
```
Comment (TomoTsuyuki) [inline classes/massactionutils.php:176]: Do we need a space between `string` and `$callbackname`?

File: classes/form/section_select_form.php [hunk 4]
```diff
@@ -100,47 +102,79 @@ public function definition() {
...
-                $radioarray[] = $mform->createElement('radio', 'targetsectionnum', '',
-                    get_string('newsection', 'block_massaction'), $targetsectionnum + 1, ['class' => 'mt-2']);
-            }
-        }
+        // Whether user can add new section.
+        $canaddnewsection = $canaddsection && ($targetsectionnum + 1) <= $targetformat->get_max_sections();
+        $this->is_allowed($canaddnewsection, $targetsectionnum + 1, $attributes);
+        $radioarray[] = $mform->createElement('radio', 'targetsectionnum', '',
+            get_string('newsection', 'block_massaction'), $targetsectionnum + 1, $attributes);
+
```
Comment (TomoTsuyuki) [inline classes/form/section_select_form.php:140]: FOUND 1 ERROR AFFECTING 1 LINE
 140 | ERROR | [x] Functions must not contain multiple empty lines in a row; found 2 empty lines

**Review by TomoTsuyuki — COMMENTED** [review]
Thanks for making PR for the issue.
It looks ok but some variable/function name are confusing.

File: classes/form/section_select_form.php (hunk 1 above)
Comment (tuanngocnguyen) [reply classes/form/section_select_form.php]: Hi @TomoTsuyuki 

Thanks for your review, I have changed it to $allowedsections, which only contains section numbers which are allowed to be chosen.

File: classes/form/section_select_form.php (hunk 2 above)
Comment (tuanngocnguyen) [reply classes/form/section_select_form.php]: I have changed this to set_radio_option

File: classes/form/section_select_form.php (hunk 4 above)
Comment (tuanngocnguyen) [reply classes/form/section_select_form.php:140]: I have removed the redundant empty line

Comment (tuanngocnguyen) [reply classes/form/section_select_form.php:140]: Thanks @TomoTsuyuki for your review

I have updated the patch

**Comment by tuanngocnguyen** [comment]
Hi @Syxton ,

Hope you are doing fine.

Would you please help with the review once you have time and let me know if there is any issue I need to address?

Regards

**Review by Syxton — COMMENTED** [review]
@tuanngocnguyen I have looked over these changes and they seem like they are good.  @PhMemmel Do you have any concerns with moving this forward?

**Comment by tuanngocnguyen** [comment]
Hi @Syxton @PhMemmel ,

Thanks for your comment.

I am happy to wait for your review.

I think the limitation should be restricted to course format.

If other plugins can implement the hooks/callbacks, we may go into a case where a course has multiple blocks with the feature. One allows to add new section, one disallows to add new section. I may lead to unpredictable behavior.
On the other hand, a course only has one format, so we will not run into this issue.

Please let me know if your opinion.


Regards

**Comment by PhMemmel** [comment]
Hi everyone,

I now took some time to come back at this and thought a little bit more about this approach:

- IMO it is not a good approach and not good style to create dependencies between plugins. Unfortunately, I could not find the place in the moodle dev docs where this is mentioned, but I'm pretty sure I've also read it there.
- I admit, these kinds of callbacks are at least loosely coupled, but I consider it as very bad style that a course format main class contains some callback functions that are not related to the course format code or moodle core apis at all, but instead are hooked in by a plugin distantly related to the course format.
- Also, block_massaction needs to contain course format related code when it is not necessary
- Callbacks are being reworked and removed from the whole moodle code and be replaced by the hooks API, so I suggest to use the hooks API for this kind of scenario. This makes these scenarios much more controlled and clean.
- The hooks API also allows other plugins to hook into block_massaction, so we do not have code for a super specific use case of special course formats, but instead we offer a hook for every plugin to use (if wanted).

Because of this I have reworked the current callback from #99 and included an alternative for this PR using a hook, see https://github.com/Syxton/moodle-block_massaction/pull/118.
It's about 95% done IMO, it basically just lacks some unit tests which test the hook functionality.

To show, how this hook can be used I also pushed a moodle branch demonstrating the hook usage for format_topics: https://github.com/moodle/moodle/compare/main...PhMemmel:moodle:Hook_example_block_massaction

@Syxton and also @tuanngocnguyen and @TomoTsuyuki: Let me know what you are thinking of this. From my point of view this would be the better approach for on the one hand addressing your (very specific) use cases and on the other hand keep the interactions with other plugins well structured and controlled.

Regarding your concerns @tuanngocnguyen about the course format having to be the only plugin controlling the section restrictrions: I designed the hook in a way so that every plugin which hooks in can disable additional sections and options, but never can add them again. So your course format callback should be good.
Also, the admin of a site can have a look in the hooks overview page in the website administration and see which hook listeners are registered, so it's easy to determine if there are other plugins hooking in. If you still do not feel well about this, it would also be possible to make the hook stoppable and to specify a priority to make sure your course format plugin will be the first.

@Syxton As my PR is more or less an API change we would have to think about how to deal with branches and releases. Also, this PR currently uses dependency injection features which is requiring moodle 4.4. If we need to be downwards compatible to moodle 4.3 this is easily reworkable however, makes it hard to unit test though.

**Comment by Syxton** [comment]
@PhMemmel I do like the hooks approach more due to the push towards hooks and their flexibility/control.  I will create a branch for our current release through 4.3, and then we can push #118 onto master (the 4.4+ branch) moving forward once completed.

**Comment by PhMemmel** [comment]
@Syxton Glad to hear and thank you very much for the feedback. Will wait for some additional feedback from @tuanngocnguyen @TomoTsuyuki, especially if my implementation addresses all of their use cases.

After that I'll finalize my PR, will finally put it against master branch which then will be our new branch for Moodle 4.4 onwards.

**Comment by tuanngocnguyen** [comment]
Hi @PhMemmel  @Syxton  ,


Thanks for your work on the issue.
The PR looks good to me. However, we are still using Moodle 4.1 and PHP 7.4, there are some incompatible issues:

> class "\core\di" does not exist in Moodle 4.1
> Read only properties are only allowed in PHP 8.1
> Constructor property is only allowed in PHP 8.0

As Moodle 4.1 is LTS, would it still be possible for our PR to be integrated into a separated branch for the moodle version ? Please let me know if you require additional work for the separated branch?

Regards

**Comment by tuanngocnguyen** [comment]
I have rebased the PR to MOODLE_401_STABLE

**Comment by Syxton** [comment]
@PhMemmel I agree.  I think that it is late in the game to be adding features to the 401 branch.  Moodle LTS general support ended in December of '23 with only security support till 2025.  I can only speak for myself that my attention is spread very thin, so I think it best to instead of adding a feature for 4.1, we focus on the long term solution with hooks and @TomoTsuyuki & @tuanngocnguyen can keep their fork with this PR until they are able to upgrade their Moodle to the hooks capable version.

**Comment by tuanngocnguyen** [comment]
Hi @PhMemmel @Syxton 

Thanks for your comments.
We will go with a fork for this PR.

Regards

**Comment by tuanngocnguyen** [comment]
Thanks all for your work.

I am closing this WR.

**Pull request closed without merging**